src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -486,23 +569,148 @@ InsertValidationResponse insertBufferedRecords(StreamingBuffer streamingBufferTo
-   *
-   * <p>Note that insertRows API does perform channel validation which might throw SFException if
-   * channel is invalidated.
-   *
-   * <p>It can also send errors {@link
-   * net.snowflake.ingest.streaming.InsertValidationResponse.InsertError} in form of response inside
-   * {@link InsertValidationResponse}
-   *
-   * @param buffer buffer to insert into snowflake
-   * @return InsertValidationResponse object sent from insertRows API.
-   */
-  private InsertValidationResponse insertRowsWithFallback(StreamingBuffer buffer) {
+  public class InsertRowsWithRetryResponse {
+    public List<InsertValidationResponse.InsertError> insertErrors;
+    public List<SinkRecord> recordWithError;
+
+    private Map<String, String> extraColumnToType;
+
+    private List<String> nonNullableColumns;
+
+    private boolean isChannelReopened;
+
+    private InsertRowsWithRetryResponse() {
+      this.insertErrors = new ArrayList<>();
+      this.recordWithError = new ArrayList<>();
+      this.isChannelReopened = false;
+    }
+
+    public boolean hasErrors() {
+      return !recordWithError.isEmpty();
+    }
+
+    public void setChannelReopened() {
+      this.isChannelReopened = true;
+    }
+
+    public boolean getChannelReopened() {
+      return this.isChannelReopened;
+    }
+
+    /** Add rows with their errors to the response */
+    public void addRecordsWithErrorsToResponse(
+        List<InsertValidationResponse.InsertError> insertErrors,
+        List<SinkRecord> insertedRecordsToBuffer) {
+      for (InsertValidationResponse.InsertError insertError : insertErrors) {
+        this.insertErrors.add(insertError);
+        this.recordWithError.add(insertedRecordsToBuffer.get((int) insertError.getRowIndex()));
+      }
+    }
+
+    /**
+     * Invoked to insert rows with extra column error into the failed buffer to retry insertion
+     *
+     * <p>Records that cannot be reinserted will be added to this response, and will be added to DLQ
+     * later on
+     *
+     * @param insertErrors errors from validation response. (Only if it has errors)
+     * @param insertedRecordsToBuffer to map {@link SinkRecord} with insertErrors
+     * @return the failed buffer, a buffer of records to be reinserted
+     */
+    private StreamingBuffer insertRecordsToFailedBuffer(
+        List<InsertValidationResponse.InsertError> insertErrors,
+        List<SinkRecord> insertedRecordsToBuffer) {
+      StreamingBuffer failedBuffer = new StreamingBuffer();
+
+      Map<String, String> schemaMap = new HashMap<>();
+      if (sfConnectorConfig
+          .get(SnowflakeSinkConnectorConfig.VALUE_CONVERTER_CONFIG_FIELD)
+          .equals(SnowflakeSinkConnectorConfig.CONFLUENT_AVRO_CONVERTER)) {
+        schemaMap =
+            SchematizationUtils.getSchemaMapForTopic(
+                topicPartition.topic(),
+                sfConnectorConfig.get(
+                    SnowflakeSinkConnectorConfig.VALUE_SCHEMA_REGISTRY_CONFIG_FIELD));
+      }
```

Comment:
do we need to get this everytime we insert to failed buffer?

Reply:
Ideally this won't happen many times, unless all the records have errors that cannot be resolved and need to be inserted to DLQ (otherwise we will error out pretty fast).

We could also cache the schema and call this function when we see new schema ID.